File: internal/app/prize_service.go
package app

import (
	"context"
	"spahtmx/internal/domain"
)

type PrizeService struct {
	repo domain.PrizeRepository
}

func NewPrizeService(r domain.PrizeRepository) *PrizeService {
	return &PrizeService{
		repo: r,
	}
}

func (s *PrizeService) GetPrizes(ctx context.Context) ([]domain.Prize, error) {
	return s.repo.GetPrizes(ctx)
}

func (s *PrizeService) GetPrize(ctx context.Context, id string) (domain.Prize, error) {
	return s.repo.GetPrize(ctx, id)
}

func (s *PrizeService) GetPrizesByYear(ctx context.Context, year string) ([]domain.Prize, error) {
	return s.repo.GetPrizesByYear(ctx, year)
}

func (s *PrizeService) GetPrizesByCategory(ctx context.Context, category string) ([]domain.Prize, error) {
	return s.repo.GetPrizesByCategory(ctx, category)
}

func (s *PrizeService) GetPrizesByCategoryAndYear(ctx context.Context, category string, year string) ([]domain.Prize, error) {
	return s.repo.GetPrizesByCategoryAndYear(ctx, category, year)
}

func (s *PrizeService) GetCategories(ctx context.Context) ([]string, error) {
	return s.repo.GetCategories(ctx)
}

func (s *PrizeService) GetYears(ctx context.Context) ([]string, error) {
	return s.repo.GetYears(ctx)
}

File: src/app/prize_service.rs
use std::sync::Arc;
use crate::domain::repositories::PrizeRepository;
use crate::domain::models::Prize;

pub struct PrizeService {
    repo: Arc<dyn PrizeRepository>,
}

impl PrizeService {
    pub fn new(repo: Arc<dyn PrizeRepository>) -> Self {
        Self { repo }
    }

    pub async fn get_prizes(&self) -> Result<Vec<Prize>, Box<dyn std::error::Error + Send + Sync>> {
        self.repo.get_prizes().await
    }

    pub async fn get_prize(&self, id: &str) -> Result<Prize, Box<dyn std::error::Error + Send + Sync>> {
        self.repo.get_prize(id).await
    }

    pub async fn get_prizes_by_year(&self, year: &str) -> Result<Vec<Prize>, Box<dyn std::error::Error + Send + Sync>> {
        self.repo.get_prizes_by_year(year).await
    }

    pub async fn get_prizes_by_category(&self, category: &str) -> Result<Vec<Prize>, Box<dyn std::error::Error + Send + Sync>> {
        self.repo.get_prizes_by_category(category).await
    }

    pub async fn get_prizes_by_category_and_year(&self, category: &str, year: &str) -> Result<Vec<Prize>, Box<dyn std::error::Error + Send + Sync>> {
        self.repo.get_prizes_by_category_and_year(category, year).await
    }

    pub async fn get_categories(&self) -> Result<Vec<String>, Box<dyn std::error::Error + Send + Sync>> {
        self.repo.get_categories().await
    }

    pub async fn get_years(&self) -> Result<Vec<String>, Box<dyn std::error::Error + Send + Sync>> {
        self.repo.get_years().await
    }
}
